[clang-tidy] Add new check bugprone-tagged-union-member-count

tigbr opened the pull request:
This patch introduces a new check to find mismatches between the number of data members in a union and the number enum values present in variant-like structures.

Variant-like types can look something like this:

```c++
struct variant {
    enum {
        tag1,
        tag2,
    } kind;
    union {
        int i;
        char c;
  } data;
};
```

The kind data member of the variant is supposed to tell which data member of the union is valid, however if there are fewer enum values than union members, then it is likely a mistake.

The opposite is not that obvious, because it might be fine to have more enum values than union data members, but for the time being I am curious how many real bugs can be caught if we give a warning regardless.

This patch also contains a heuristic where we try to guess whether the last enum constant is actually supposed to be a tag value for the variant or whether it is just holding how many enum constants have been created.

I am still collecting data on how many bugs this check can catch on open source software and I will update the pull-request with the results.

tigbr committed:
[clang-tidy] Add new check bugprone-tagged-union-member-count

This patch introduces a new check to find mismatches between the number
of data members in a union and the number enum values present in
variant-like structures.

Variant-like types can look something like this:

```c++
struct variant {
    enum {
        tag1,
        tag2,
    } kind;
    union {
        int i;
        char c;
  } data;
};
```

The kind data member of the variant is supposed to tell which data
member of the union is valid, however if there are fewer enum values
than union members, then it is likely a mistake.

The opposite is not that obvious, because it might be fine to have more
enum values than union data members, but for the time being I am curious
how many real bugs can be caught if we give a warning regardless.

This patch also contains a heuristic where we try to guess whether the
last enum constant is actually supposed to be a tag value for the
variant or whether it is just holding how many enum constants have been
created.

I am still collecting data on how many bugs this check can catch on open
source software and I will update the pull-request with the results.

EugeneZelenko commented on clang-tools-extra/clang-tidy/bugprone/TaggedUnionMemberCountCheck.cpp:
```diff
@@ -0,0 +1,125 @@
+#include "clang/AST/ASTContext.h"
+#include "clang/AST/Expr.h"
+#include "clang/ASTMatchers/ASTMatchFinder.h"
+
```
Excessive newline.

EugeneZelenko commented on clang-tools-extra/clang-tidy/bugprone/TaggedUnionMemberCountCheck.cpp:
```diff
@@ -0,0 +1,125 @@
+  int tags = 0;
+  int unions = 0;
+  for (const FieldDecl *r : rec->fields()) {
+    TypeSourceInfo *info = r->getTypeSourceInfo();
```
Should be `const`.

PiotrZSL replied:
per codding standard local variables should be CammelCase

EugeneZelenko commented on clang-tools-extra/docs/clang-tidy/checks/bugprone/tagged-union-member-count.rst:
```diff
@@ -0,0 +1,66 @@
+    tags_count, // is 3
+  };
+
+The checker detects this usage pattern heuristically and does not include
```
```suggestion
The check detects this usage pattern heuristically and does not include
```

EugeneZelenko commented on clang-tools-extra/test/clang-tidy/checkers/bugprone/tagged-union-member-count.cpp:
```diff
@@ -0,0 +1,181 @@
+// RUN: %check_clang_tidy %s bugprone-tagged-union-member-count %t
+
+enum tags3 {
+	tags3_1,
```
Will be good idea to run Clang-format over this file.

EugeneZelenko commented on clang-tools-extra/test/clang-tidy/checkers/bugprone/tagged-union-member-count.cpp:
```diff
@@ -0,0 +1,181 @@
+		} complex;
+	} data;
+};
+
```
Excessive newline.

PiotrZSL replied:
add tests with typedefs, where enum type is behind typedef, or union is behind typedef, this may require you to get "CanonicalType" of enum and record.

isuckatcs replied:
Please add tests with templates, and enums with custom underlying types. In C++, we also have `enum class`, which could also be used as a tag type, so please add test cases for that too.

PiotrZSL commented on clang-tools-extra/clang-tidy/bugprone/TaggedUnionMemberCountCheck.cpp:
```diff
@@ -0,0 +1,125 @@
+
+  // FIXME: Maybe a emit a note when a counter enum constant was found.
+  if (unionMemberCount > tagCount) {
+    diag(root->getLocation(), "Tagged union has more data members than tags! "
+                              "Data members: %0 Tags: %1")
```
should be more a:
"tagged union has more data members (%0) than tags (%1)"
some suggestion what to do would be also nice.

PiotrZSL commented on clang-tools-extra/clang-tidy/bugprone/TaggedUnionMemberCountCheck.cpp:
```diff
@@ -0,0 +1,125 @@
+    if (enumValue < minTagValue)
+      minTagValue = enumValue;
+
+    if (enumName.ends_with_insensitive("count")) {
```
what if this won't be count, but will be "last", or "size", or something. Add config option for this and use regexp for matching.

PiotrZSL commented on clang-tools-extra/clang-tidy/bugprone/TaggedUnionMemberCountCheck.h:
```diff
@@ -0,0 +1,31 @@
+  TaggedUnionMemberCountCheck(StringRef Name, ClangTidyContext *Context)
+      : ClangTidyCheck(Name, Context) {}
+  void registerMatchers(ast_matchers::MatchFinder *Finder) override;
+  void check(const ast_matchers::MatchFinder::MatchResult &Result) override;
```
use IgnoreUnlessSpelledInSource to exclude implicit code, look into other checks.

tigbr replied:
This option looks interesting, however it breaks some of the intended matches, for example [the test case in C++ when the union field is anonymous](https://github.com/tigbr/llvm-project/blob/be3ec1a4e9ac3309da8f2037334de68bd478883f/clang-tools-extra/test/clang-tidy/checkers/bugprone/tagged-union-member-count.cpp#L176).

PiotrZSL commented on clang-tools-extra/docs/ReleaseNotes.rst:
```diff
@@ -129,6 +129,12 @@ New checks
   Replaces certain conditional statements with equivalent calls to
   ``std::min`` or ``std::max``.
 
+- New :doc:`bugprone-tagged-union-member-count
```
move to line 119, keep in alphabetical order

PiotrZSL commented on clang-tools-extra/docs/ReleaseNotes.rst:
```diff
@@ -129,6 +129,12 @@ New checks
+- New :doc:`bugprone-tagged-union-member-count
+  <clang-tidy/checks/bugprone/tagged-union-member-count>` check.
+
+  Warns about suspicious looking tagged unions where the number of enum
+  constants and the number of union data members are not equal.
```
class documentation, release notes entry, and first sentence in documentation should be in sync.

PiotrZSL commented on clang-tools-extra/docs/clang-tidy/checks/list.rst:
```diff
@@ -142,6 +142,7 @@ Clang-Tidy Checks
    :doc:`bugprone-suspicious-stringview-data-usage <bugprone/suspicious-stringview-data-usage>`,
    :doc:`bugprone-swapped-arguments <bugprone/swapped-arguments>`, "Yes"
    :doc:`bugprone-switch-missing-default-case <bugprone/switch-missing-default-case>`,
+   :doc:`bugprone-tagged-union-member-count <bugprone/tagged-union-member-count>`, "Yes"
```
check does not provide fixes, so remove `"Yes"`

PiotrZSL commented on clang-tools-extra/test/clang-tidy/checkers/bugprone/tagged-union-member-count.cpp:
```diff
@@ -0,0 +1,181 @@
+};
+
+struct taggedunion5 { // CHECK-MESSAGES: :[[@LINE]]:8: warning: Tagged union has more data members than tags! Data members: 4 Tags: 3 [bugprone-tagged-union-member-count]
+	enum tags3 tag;
```
run clang-format on this file (you may need to move it somewere to overcome clang-format exclude of test directory, just to format it and bring it back again)

PiotrZSL commented on clang-tools-extra/test/clang-tidy/checkers/bugprone/tagged-union-member-count.cpp:
```diff
@@ -0,0 +1,181 @@
+	union union3 dataA;
+};
+
+struct taggedunion1 { // CHECK-MESSAGES: :[[@LINE]]:8: warning: Tagged union has more data members than tags! Data members: 4 Tags: 3 [bugprone-tagged-union-member-count]
```
put those logs in one line up, just simply use `[[@LINE+1]]`

PiotrZSL reviewed (changes requested):
This check may not have to many users, but that's fine.
I'm 100% sure that I seen such structures already.

As for code changes, please fix pointed out nits and after that checkout probably could be merged.

isuckatcs commented on clang-tools-extra/clang-tidy/bugprone/TaggedUnionMemberCountCheck.cpp:
```diff
@@ -0,0 +1,125 @@
+  int64_t tagCount = getNumberOfValidEnumValues(tagMatch);
+
+  // FIXME: Maybe a emit a note when a counter enum constant was found.
+  if (unionMemberCount > tagCount) {
```
This can lead to false negatives. These are not possible to detect for the same reason the counter enumerator isn't either, but it worths noting it.

E.g.:
```c++
struct TU {
  enum {
    UNASSIGNED,
    INT,
    FLOAT,
  } tags;

  union {
    int i;
    float f;
    long long ll;
  } data;
};
```

On the other hand, I guess if the condition is `true`, the diagnostics should always be reported.

isuckatcs commented on clang-tools-extra/clang-tidy/bugprone/TaggedUnionMemberCountCheck.cpp:
```diff
@@ -0,0 +1,125 @@
+    diag(root->getLocation(), "Tagged union has more data members than tags! "
+                              "Data members: %0 Tags: %1")
+        << unionMemberCount << tagCount;
+  } else if (unionMemberCount < tagCount) {
```
This here might not always be correct.

```c++
struct TU {
  enum {
    UNASSIGNED,
    INT,
    FLOAT,
  } tags;

  union {
    int i;
    float f;
  } data;
};
```

tigbr replied:
This diagnostic is emitted only when the `StrictMode` option to `true`, which is disabled by default precisely because of these cases.

isuckatcs commented on clang-tools-extra/docs/clang-tidy/checks/bugprone/tagged-union-member-count.rst:
```diff
@@ -0,0 +1,66 @@
+Gives warnings for tagged unions, where the number of tags is
+different from the number of data members inside the union.
+
+A struct or a class is considered to be a tagged union if it has
```
The matcher only matches `struct` IIRC.

isuckatcs commented on clang-tools-extra/docs/clang-tidy/checks/bugprone/tagged-union-member-count.rst:
```diff
@@ -0,0 +1,66 @@
+    tag2,
+  };
+
+  struct taggedunion { // warning: Tagged union has more data members than tags! Data members: 3 Tags: 2 [bugprone-tagged-union-member-count]
```
```suggestion
  struct TaggedUnion { // warning: Tagged union has more data members than tags! Data members: 3 Tags: 2 [bugprone-tagged-union-member-count]
```

Please use the same naming convention in the examples, that is used in source code.

isuckatcs commented on clang-tools-extra/docs/clang-tidy/checks/bugprone/tagged-union-member-count.rst:
```diff
@@ -0,0 +1,66 @@
+    tag4,
+  };
+  
+  struct taggedunion { // warning: Tagged union has more fewer members than tags! Data members: 3 Tags: 4 [bugprone-tagged-union-member-count]
```
```suggestion
  struct taggedunion { // warning: Tagged union has fewer data members than tags! Data members: 3 Tags: 4 [bugprone-tagged-union-member-count]
```

Please keep this in sync with the actual diagnostics.

isuckatcs commented on clang-tools-extra/test/clang-tidy/checkers/bugprone/tagged-union-member-count.cpp:
```diff
@@ -0,0 +1,181 @@
+};
+
+// It is not obvious which enum is the tag for the union.
+struct taggedunion2 { // No warnings expected.
```
Please pay attention to the naming. This is an invalid testcase, so it should be called `MaybeTaggedUnion`, or something similar. Similarly for the other `struct`s, the naming should reflect what we attempt to test there, from `taggedunionN` it is not obvious.

isuckatcs commented on clang-tools-extra/test/clang-tidy/checkers/bugprone/tagged-union-member-count.cpp:
```diff
@@ -0,0 +1,181 @@
+	node_type_loop,
+	node_type_branch,
+	node_type_function,
+	node_type_count,
```
I might design my own language, where `count` can also be an AST node, then we report a false positive, right?

tigbr replied:
Right, the heuristic can be disabled now in the configuration.

isuckatcs commented on clang-tools-extra/test/clang-tidy/checkers/bugprone/tagged-union-member-count.cpp:
```diff
@@ -0,0 +1,181 @@
+};
+
+// Without the counter enum constant the number of tags
+// and the number data members are equal.
```
```suggestion
// and the number of data members are equal.
```

isuckatcs commented on clang-tools-extra/test/clang-tidy/checkers/bugprone/tagged-union-member-count.cpp:
```diff
@@ -0,0 +1,181 @@
+	union union3 data;
+};
+
+// Technically this means that every enum value is defined from 0-256 and therefore a warning is given.
+enum mycolor {
+	mycolor_black = 0x00,
+	mycolor_gray  = 0xcc,
+	mycolor_white = 0xff,
+};
```
I think this is incorrect here. The number of enumerators is `3`. AFAIK nothing in a tagged union requires the tag enums to have values increasing from `0`.

Even if the values are `0`, `1` and `2`, the underlying type can represent more values than these 3, and we don't say that we have `INT_MAX - INT_MIN + 1` tags, right?

isuckatcs commented on clang-tools-extra/test/clang-tidy/checkers/bugprone/tagged-union-member-count.cpp:
```diff
@@ -0,0 +1,181 @@
+// RUN: %check_clang_tidy %s bugprone-tagged-union-member-count %t
```
The default config is `-std=c++11-or-later`, but this check is not limited to that only, so we want to run it with `-std=c++98-or-later`.

The check also supports C and Objective-C, so please add `.c` and `.m` test files and create tests for those languages too. 

If you don't want to support any of these languages, please add `bool isLanguageVersionSupported(const LangOptions &LangOpts) const override;` to your check and configure the supported languages.

isuckatcs reviewed (changes requested):
I like the motivation behind the check, but I would also like to see more testing and some formatting to be fixed.

tigbr committed:
pull request fixes

whisperity commented on clang-tools-extra/clang-tidy/bugprone/TaggedUnionMemberCountCheck.h:
```diff
@@ -0,0 +1,37 @@
+  void check(const ast_matchers::MatchFinder::MatchResult &Result) override;
+
+private:
+  bool EnumCounterHeuristicIsEnabled;
+  StringRef EnumCounterSuffix;
+  bool StrictMode;
```
If these are the values to store the check config options the user specified, these should be `const` and not modified during analysis.

whisperity commented on clang-tools-extra/docs/clang-tidy/checks/bugprone/tagged-union-member-count.rst:
```diff
@@ -0,0 +1,216 @@
+    tag2,
+  };
+
+  struct taggedUnion { // warning: Tagged union has more data members (3) than tags (2)! [bugprone-tagged-union-member-count]
```
```suggestion
  struct taggedUnion { // warning: Tagged union has more data members (3) than tags (2)
```

whisperity commented on clang-tools-extra/docs/clang-tidy/checks/bugprone/tagged-union-member-count.rst:
```diff
@@ -0,0 +1,216 @@
+    tag1, // is 0
+    tag2, // is 1
+    tag3, // is 2
+    tags_count, // is 3
```
You should extend the patch with supporting a **prefix** for the tag counting enum constant. In various projects it is easy to observe enums like `LAST_Foo`.

whisperity commented on clang-tools-extra/clang-tidy/bugprone/TaggedUnionMemberCountCheck.h:
```diff
@@ -0,0 +1,37 @@
+
+private:
+  bool EnumCounterHeuristicIsEnabled;
+  StringRef EnumCounterSuffix;
```
Perhaps this should be a `vector<string>` or something of sorts. Using `Foo;Bar;Baz` as the input from the user, you can use something called `splitOptions` under `utils`. It would allow to support various styles at the same time, as some enums might be called `_Count` but others `Last_`.

whisperity commented on clang-tools-extra/test/clang-tidy/checkers/bugprone/tagged-union-member-count-strictmode.cpp:
```diff
@@ -0,0 +1,150 @@
+	} data;
+};
+
+// CHECK-MESSAGES: :[[@LINE+1]]:8: warning: Tagged union has fewer data members (3) than tags (4)! [bugprone-tagged-union-member-count]
```
```suggestion
// CHECK-MESSAGES: :[[@LINE+1]]:8: warning: Tagged union has fewer data members (3) than tags (4)
```

There is no need to repeat the full message in every line, as the check is applied as a substring and regular expression match, it is enough to type out the prefix only until the last dynamically generated part.

tigbr committed:
Major iteration based on review comments

Feature changes:

* Added C++ classes to matcher
* Added multiple prefix/suffix option support to the counting enum heuristic

Implementation changes:

* Distinct enum values are counted with a set instead of the previous max - min + 1 approach
* Use APSInt to avoid truncation or overflow

Testing changes:

* Tests use more descriptive names
* Removed redundant suffix from test case messages
* Separated test files for C, C++, Objective-C and Objective-C++
* C++ tests are ran with the -std=c98-or-later flag
* C++ tests are extended with more language feature test cases
* Created tests for the Prefix/suffix feature of the counting enum heuristic
* StrictMode and heuristic enabling/disabling is more thoroughly tested

Also made plenty of small improvements as well.

tigbr commented:
Thank you everyone for your suggestions and observations, they were most helpful. I listed the important changes topically in the message of the latest commit, I hope it will be of help during review.

EugeneZelenko commented:
Please resolve comments that already fixed.

tigbr commented on clang-tools-extra/test/clang-tidy/checkers/bugprone/tagged-union-member-count.m:
A @isuckatcs specifically requested [testing C and Objective-C or specifying in the check that those languages are not supported](https://github.com/llvm/llvm-project/pull/89925#discussion_r1578557769).

> The default config is `-std=c++11-or-later`, but this check is not limited to that only, so we want to run it with `-std=c++98-or-later`.
> 
> The check also supports C and Objective-C, so please add `.c` and `.m` test files and create tests for those languages too.
> 
> If you don't want to support any of these languages, please add `bool isLanguageVersionSupported(const LangOptions &LangOpts) const override;` to your check and configure the supported languages.

C-style tagged unions are supported in Objective-C and Objective-C++ so why exclude these languages? If we support them, then they should be tested.

isuckatcs replied:
> If we support them, then they should be tested.

100% agreed

tigbr committed:
Next batch of review comment fixes

Implementation:

* Removed redundant allOf from matcher.
* Removed unused llvm::enumerate from loop.
* Changed static_cast to llvm::dyn_cast_or_null.
* Update diagnostic message so that is starts with lower case.

Documentation:

* Rephrased confusing parts of the documentation.
* Updated the diagnostic messages in the documentation examples to start with lowercase.
* Fixed the example for the prefixes in the documentation.

New configuration diagnostic:

Implemented a diagnostic warning for the configuration of the counting enum heuristic.
When the heuristic is disabled and a prefix or a suffix is explicitly set
in the configuration file a warning is emitted.

Testing:

* Created test case for macros.
* Created test case for templates where the union and tag are not templated,
  but some other data member is.
* Moved enum counter heuristic tests from .c and .m files into its
  deticated test file.
* Created test for the new configuration diagnostic.
* Updated tagged-union-member-count-counting-enum-heuristic-is-disabled.cpp so
  that the new configuration diagnostic does not trigger on this file

Style and formatting:

* Removed braces from single statement if statements.
* Changed a couple variables to use the CamelCase naming style.
* Changed diagnostic message to start with lower case.
* Moved the option name string constants to the .cpp file.
* Renamed StrictModeIsEnabled option to just StrictMode.
* Use true and false instead of 1 and 0 in the tests of the configuration options
* Renamed CountingEnumHeuristicIsEnabled to EnableCountingEnumHeuristic (code, docs, tests).
* Keep text in 80 column limit in docs .rst file.

tigbr commented on clang-tools-extra/docs/clang-tidy/checks/bugprone/tagged-union-member-count.rst:
```diff
@@ -0,0 +1,217 @@
+
+When defined, the check will use the list of the semicolon separated strings
+in CountingEnumPrefixes or CountingEnumSuffixes for the identification of possible counting enum constants.
+These options do not alter the check's behavior when :option:`CountingEnumHeuristicIsEnabled` is set to false.
```
I don't think this is too complicated to implement. I can separate the default value and the value given by the user by calling has_value on the `std::optional` returned by the version of `OptionsView::get` that has only one parameter.

tigbr committed:
Remove outdated TODO comment.

tigbr committed:
Emit a note, when the final enum count is modified because of the
counting enum heuristic.

tigbr committed:
Fix bug that StrictMode was enabled by default.

whisperity commented on clang-tools-extra/clang-tidy/bugprone/TaggedUnionMemberCountCheck.cpp:
```diff
@@ -0,0 +1,214 @@
+         llvm::count_if(Rec->fields(), isEnum) > 1;
+}
+
+static bool signEquals(const llvm::APSInt &A, const llvm::APSInt &B) {
+  return (A.isNegative() && B.isNegative()) ||
+         (A.isStrictlyPositive() && B.isStrictlyPositive()) ||
+         (A.isZero() && B.isZero());
+}
```
**If** the following is equivalent (I would guess `isNegative`/`isStrictlyPositive`/`isZero` are all mutually exclusive properties) then perhaps expressing the query like this would better reflect that we are checking for some *equality* here:

```suggestion
static bool signEquals(const llvm::APSInt &A, const llvm::APSInt &B) {
  return (A.isNegative() == B.isNegative()) &&
         (A.isStrictlyPositive() == B.isStrictlyPositive()) &&
         (A.isZero() == B.isZero());
}
```

-----

Also, I am **very** surprised that `APInt`/`APSInt` does not have the facility for querying the [sign function](https://en.wikipedia.org/wiki/Sign_function) as a single value. Tangentially, perhaps you could add that in a separate patch with a few tests, a function that returns a 3-way enum `NEGATIVE, ZERO, POSITIVE`, and then this function simplifies to `A.getSign() == B.getSign()`, which is the most trivial expansion of the function's name.

whisperity commented on clang-tools-extra/clang-tidy/bugprone/TaggedUnionMemberCountCheck.cpp:
```diff
@@ -0,0 +1,182 @@
+  const auto *UnionField = Result.Nodes.getNodeAs<FieldDecl>("union");
+  const auto *TagField = Result.Nodes.getNodeAs<FieldDecl>("tags");
+
+  // The matcher can only narrow down the type to recordType()
+  if (!isUnion(UnionField))
+    return;
+
+  if (hasMultipleUnionsOrEnums(Root))
+    return;
```
The benefit of customised matchers with `AST_MATCHER` or `AST_MATCHER_P` instead of an imperative function call is that if we can sort out the non-matching cases declaratively, we spare ourselves the cost of building the `Result` object and firing the callback. (Also, there are *some* memoisation logics in the AST Matcher library, but that is only a theoretical benefit here, as this check only calls back each record once. The memoisation would benefit for checks where complex expressions are investigated.)

whisperity commented on clang-tools-extra/clang-tidy/bugprone/TaggedUnionMemberCountCheck.cpp:
```diff
@@ -0,0 +1,214 @@
+bool TaggedUnionMemberCountCheck::isCountingEnumLikeName(
+    StringRef Name) const noexcept {
+  if (llvm::any_of(ParsedCountingEnumPrefixes,
+                   [&Name](const StringRef &Prefix) -> bool {
```
```suggestion
                   [Name](StringRef Prefix) -> bool {
```

(Nit: `StringRef`s are glorified pointers, they are designed to be cheap to copy.)

whisperity commented on clang-tools-extra/clang-tidy/bugprone/TaggedUnionMemberCountCheck.cpp:
```diff
@@ -0,0 +1,214 @@
+  return false;
+}
+
+size_t TaggedUnionMemberCountCheck::getNumberOfValidEnumValues(
```
```suggestion
std::size_t TaggedUnionMemberCountCheck::getNumberOfValidEnumValues(
```

(`size_t` being defined is not guaranteed in a portable way.)

whisperity commented on clang-tools-extra/clang-tidy/bugprone/TaggedUnionMemberCountCheck.cpp:
```diff
@@ -0,0 +1,214 @@
+    const EnumDecl *Ed) noexcept {
+  bool FoundMax = false;
+  llvm::APSInt MaxTagValue;
+  llvm::SmallSet<llvm::APSInt, 32> EnumValues;
```
(Nit: Perhaps `32` elements is a rather large number for the general case, but this is just a theory.)

tigbr replied:
I tested the check on several open source projects and the number of enum values in the warnings were `[3, 3, 4, 4, 6, 8, 9, 10, 11, 32, 33, 46]`. The average of these is about 14, so maybe the value of 16 covers most cases without extra allocation.

whisperity commented on clang-tools-extra/clang-tidy/bugprone/TaggedUnionMemberCountCheck.cpp:
```diff
@@ -0,0 +1,214 @@
+        CeIsLast = true;
+        CeValue = Val;
+        CeCount += 1;
+        CountingEnumConstantDecl = Enumerator;
```
This `CountingEnumConstantDecl` is a quasi-global symbol as the member of the check class. Also, it is not cleared after the last diagnostic is printed, and we only rely on the class never getting used again in order to not have something dangling here.

Perhaps it would be better if the function returned `std::pair<std::size_t, const Enum...*>`, or a `const Enum...*&` output parameter was taken as a side effect.

whisperity commented on clang-tools-extra/clang-tidy/bugprone/TaggedUnionMemberCountCheck.cpp:
```diff
@@ -0,0 +1,214 @@
+
+void TaggedUnionMemberCountCheck::check(
+    const MatchFinder::MatchResult &Result) {
+  const auto *Root = Result.Nodes.getNodeAs<RecordDecl>("root");
+  const auto *UnionField = Result.Nodes.getNodeAs<FieldDecl>("union");
+  const auto *TagField = Result.Nodes.getNodeAs<FieldDecl>("tags");
+
```
```suggestion
  const auto *Root = Result.Nodes.getNodeAs<RecordDecl>(RootNode);
  const auto *UnionField = Result.Nodes.getNodeAs<FieldDecl>(UnionNode);
  const auto *TagField = Result.Nodes.getNodeAs<FieldDecl>(EnumNode);
  
  assert(Root && UnionField && TagField);
```

whisperity commented on clang-tools-extra/clang-tidy/bugprone/TaggedUnionMemberCountCheck.cpp:
```diff
@@ -0,0 +1,214 @@
+  if (hasMultipleUnionsOrEnums(Root))
+    return;
+
+  const auto *UnionDef =
+      UnionField->getType().getCanonicalType().getTypePtr()->getAsRecordDecl();
+  const auto *EnumDef = llvm::dyn_cast_or_null<EnumDecl>(
+      TagField->getType().getCanonicalType().getTypePtr()->getAsTagDecl());
+
+  if (EnumDef) {
```
```suggestion
  const auto *UnionDef =
      UnionField->getType().getCanonicalType().getTypePtr()->getAsRecordDecl();
  assert(UnionDef);
  const auto *EnumDef = llvm::dyn_cast_or_null<EnumDecl>(
      TagField->getType().getCanonicalType().getTypePtr()->getAsTagDecl());

  if (!EnumDef)
    return;
```

(And then pull the inner block of the current `if` one indentation out.)

whisperity commented on clang-tools-extra/docs/clang-tidy/checks/bugprone/tagged-union-member-count.rst:
```diff
@@ -0,0 +1,241 @@
+any number of other data members that are neither unions or enums.
+
+Example
+=======
```
```suggestion
-------
```

(Although the example(s) are usually not individual heading levels in the Tidy documentation.)

whisperity commented on clang-tools-extra/docs/clang-tidy/checks/bugprone/tagged-union-member-count.rst:
```diff
@@ -0,0 +1,241 @@
+  };
+
+Counting enum constant heuristic
+================================
```
```suggestion
--------------------------------
```

(...etc.)

whisperity commented on clang-tools-extra/docs/clang-tidy/checks/bugprone/tagged-union-member-count.rst:
```diff
@@ -0,0 +1,241 @@
+Counting enum constant heuristic
+================================
+
+Sometimes the last enum constant in an enum is used to keep track of how many
+enum constants have been declared. For an illustration:
+
+.. code-block:: c++
+
+  enum TagWithCounter {
+    Tag1, // is 0
+    Tag2, // is 1
+    Tag3, // is 2
+    TagCount, // is 3
+  };
```
```suggestion
Sometimes the last enum constant in an enum is used to keep track of how many
enum constants have been declared, or to explicitly "point to" the last declared valid enum constant. For an illustration:

.. code-block:: c++

  enum TagWithCounter {
    Tag1, // is 0
    Tag2, // is 1
    Tag3, // is 2
    TagCount, // is 3
  };
  
  enum TagWithLast {
    Tag1 = 0,
    Tag2 = 1,
    Tag3 = 2,
    LastTag = 2,
  };
```

(and check if you have proper tests and behaviour for these.)

whisperity commented on clang-tools-extra/docs/clang-tidy/checks/bugprone/tagged-union-member-count.rst:
```diff
@@ -0,0 +1,241 @@
+case insensitively.
+
+Options
+=======
```
```suggestion
-------
```

whisperity reviewed (changes requested):
I found a very convoluted and somewhat silly edge case where I was able to defeat the matchers to make them produce a match in a position where it is clearly not intended. So the matchers will need to be given some thought, my idea is to have them constricted in a way to report only into non-implicit code, but perhaps this needs some more elaboration.

Lambdas implicitly compile down to an unnamed `CXXRecordDecl`, and if you have captures in a lambda, they become (unnamed) fields of this class.

https://godbolt.org/z/rvfY5K18T

```cpp
enum E { A };
union U {
    int A;
};

int main() {
    enum E e;
    union U u;
    auto L = [e, u] () {};
}
```

```
Match #1:

<source>:9:14: note: "root" binds here
    9 |     auto L = [e, u] () {};
      |              ^
<source>:9:15: note: "tags" binds here
    9 |     auto L = [e, u] () {};
      |               ^
<source>:9:18: note: "union" binds here
    9 |     auto L = [e, u] () {};
      |                  ^
```

tigbr committed:
Implemented the suggestions of latest review comments.

isuckatcs commented on clang-tools-extra/clang-tidy/bugprone/TaggedUnionMemberCountCheck.cpp:
```diff
@@ -98,22 +124,21 @@ static bool greaterBySign(const llvm::APSInt &A, const llvm::APSInt &B) {
 
-size_t TaggedUnionMemberCountCheck::getNumberOfValidEnumValues(
-    const EnumDecl *Ed) noexcept {
+std::size_t TaggedUnionMemberCountCheck::getNumberOfValidEnumValues(
+    const EnumDecl *ED,
+    const EnumConstantDecl *&OutCountingEnumConstantDecl) {
```
```suggestion
std::pair<std::size_t, const EnumConstantDecl *> TaggedUnionMemberCountCheck::getNumberOfValidEnumValues(
    const EnumDecl *ED) {
```

isuckatcs commented on clang-tools-extra/clang-tidy/bugprone/TaggedUnionMemberCountCheck.cpp:
```diff
@@ -132,11 +157,11 @@ size_t TaggedUnionMemberCountCheck::getNumberOfValidEnumValues(
   // 3. The value of the counting enum constant is the largest out of every enum
   //    constant.
```
So, if I understand it correctly, what we really care about is whether the value of the last enumerator equals to the number of the other enumerators.

E.g.:
```c++
enum tag_with_counter {
  tag1 = 100,
  tag2 = 200,
  tag_count = 2
};
```
In this example `tag_count` is still a counter, but point 3. doesn't apply, so the heuristics doesn't find it.

```c++
enum tag_with_counter {
  tag1 = 2,
  tag2 = 3,
  tag_count = 2
};
```
In this other example, the last enumerator is still a counter, but point 1. doesn't apply (?).

```c++
enum tag_with_counter {
  tag1,
  tag2,
  tag3
};
```
In this last example however all points apply, so the heuristics detects the last enumerator as a counter, when it's not one.

If you really want this heuristics, I suggest replacing this function with a simple check for the value of the last enumerator being equal to the number of the other enumerators. You use the name of the enumerator to determine whether it's a counter anyway.

tigbr replied:
I agree that this part was overcomplicated. I changed it to just checking the name and the value of the last enum constant.

isuckatcs commented on clang-tools-extra/clang-tidy/bugprone/TaggedUnionMemberCountCheck.cpp:
```diff
@@ -153,61 +178,57 @@ size_t TaggedUnionMemberCountCheck::getNumberOfValidEnumValues(
+  const auto *UnionField =
+      Result.Nodes.getNodeAs<FieldDecl>(UnionMatchBindName);
+  const auto *TagField = Result.Nodes.getNodeAs<FieldDecl>(TagMatchBindName);
+  assert(Root && UnionField && TagField);
```
How about adding a return if one of these is missing, so we don't crash in release mode? Also it's a good idea to separate the assertions, so that we know which of the 3 conditions fails on a crash.

isuckatcs commented on clang-tools-extra/clang-tidy/bugprone/TaggedUnionMemberCountCheck.cpp:
```diff
@@ -153,61 +178,57 @@ size_t TaggedUnionMemberCountCheck::getNumberOfValidEnumValues(
-  const auto *EnumDef = llvm::dyn_cast_or_null<EnumDecl>(
+  const auto *EnumDef = llvm::dyn_cast<EnumDecl>(
       TagField->getType().getCanonicalType().getTypePtr()->getAsTagDecl());
+  assert(UnionDef && EnumDef && "Declarations missing!");
```
Same here.

isuckatcs commented on clang-tools-extra/docs/clang-tidy/checks/bugprone/tagged-union-member-count.rst:
```diff
@@ -30,37 +28,74 @@ Example
 
-Sometimes the last enum constant in an enum is used to keep track of how many
-enum constants have been declared. For an illustration:
+For an illsutration:
```
```suggestion
For an illustration:
```

isuckatcs commented on clang-tools-extra/docs/clang-tidy/checks/bugprone/tagged-union-member-count.rst:
```diff
@@ -30,37 +28,74 @@ Example
+------------------------------
+
+The main complicating factor when counting the number of enum constants is that
+some of them might be auxiliary values that purposely don't have a corresponding union
```
```suggestion
some of them might be auxiliary values that purposefully don't have a corresponding union
```

isuckatcs commented on clang-tools-extra/docs/clang-tidy/checks/bugprone/tagged-union-member-count.rst:
```diff
@@ -30,37 +28,74 @@ Example
-"counting" enum constant will intentionally not have a corresponding union
-data member.
+The check counts the number of distinct values among the enum constants and not the enum
+constants themselves. This way enum constants that are essentially just aliases of other
```
```suggestion
constants themselves. This way the enum constants that are essentially just aliases of other
```

tigbr committed:
Implement the requested changes of isuckatcs.

tigbr commented on clang-tools-extra/clang-tidy/bugprone/TaggedUnionMemberCountCheck.cpp:
```diff
@@ -16,39 +16,74 @@ using namespace clang::ast_matchers;
+static constexpr llvm::StringLiteral TagMatchBindName = "tags";
+
+namespace {
+AST_MATCHER(FieldDecl, isUnion) {
+  const Type *T = Node.getType().getCanonicalType().getTypePtr();
+  assert(T);
+  return T->isUnionType();
+}
+
+AST_MATCHER(FieldDecl, isEnum) {
+  const Type *T = Node.getType().getCanonicalType().getTypePtr();
+  assert(T);
+  return T->isEnumeralType();
+}
```
I don't think these are the same as the ones in the [AST Matcher Reference](https://clang.llvm.org/docs/LibASTMatchersReference.html). In need these matchers to work with `FieldDecls` while those work on `TagDecls`.

whisperity replied:
@tigbr How about these matchers? https://godbolt.org/z/Mh3nEo8YY

```cpp
match fieldDecl(
    hasType(qualType(hasCanonicalType(
        recordType(hasDeclaration(
            recordDecl(isUnion()).bind("union-decl")
        ))))
    )
).bind("union-field")

match fieldDecl(
    hasType(qualType(hasCanonicalType(
        enumType(hasDeclaration(
            enumDecl().bind("enum-decl")
        )
        )))
    )
).bind("enum-field")
```

tigbr replied:
I was able to remove these AST_MATCHER macros by using the matchers that @whisperity has suggested.

whisperity commented on clang-tools-extra/clang-tidy/bugprone/TaggedUnionMemberCountCheck.cpp:
```diff
@@ -0,0 +1,192 @@
+    LastEnumConstant = Enumerator;
+  }
+
+  if (EnableCountingEnumHeuristic && LastEnumConstant) {
+    if (isCountingEnumLikeName(LastEnumConstant->getName())) {
+      if (LastEnumConstant->getInitVal() == (EnumValues.size() - 1)) {
+        return {EnumValues.size() - 1, LastEnumConstant};
+      }
+    }
+  }
```
```suggestion
  if (EnableCountingEnumHeuristic && LastEnumConstant &&
      isCountingEnumLikeName(LastEnumConstant->getName()) &&
      (LastEnumConstant->getInitVal() == EnumValues.size() - 1))
    return {EnumValues.size() - 1, LastEnumConstant};
```

tigbr committed:
Replace AST_MATCHER macros and fix coding style in a few places.

tigbr committed:
Change starting size of SmallSet from 32 to 16.

tigbr commented on clang-tools-extra/clang-tidy/bugprone/TaggedUnionMemberCountCheck.h:
```diff
@@ -0,0 +1,45 @@
+  const StringRef RawCountingEnumPrefixes;
+  const StringRef RawCountingEnumSuffixes;
+  std::vector<StringRef> ParsedCountingEnumPrefixes;
+  std::vector<StringRef> ParsedCountingEnumSuffixes;
```
The StringRefs in the result vector point to the StringRef that is passed to parseStringList. The input into parseStringList is obtained from the `OptionsView Options;` data member of my ClangTidyCheck object. So I think these are valid while my check object is alive.

[parseStringList](https://clang.llvm.org/extra//doxygen/OptionsUtils_8cpp_source.html#l00016) [ClangTidyCheck](https://clang.llvm.org/extra/doxygen/classclang_1_1tidy_1_1ClangTidyCheck.html)

Szelethus commented:
My say doesn't quite have the same weight here are many others, but I see no problems with this patch after so many rounds of reviews -- would it be okay to merge?

EugeneZelenko commented on clang-tools-extra/docs/clang-tidy/checks/bugprone/tagged-union-member-count.rst:
```diff
@@ -0,0 +1,275 @@
+
+This option is enabled by default.
+
+When :option:`EnableCountingEnumHeuristic` is false:
```
```suggestion
When :option:`EnableCountingEnumHeuristic` is `false`:
```

EugeneZelenko commented on clang-tools-extra/docs/clang-tidy/checks/bugprone/tagged-union-member-count.rst:
```diff
@@ -0,0 +1,275 @@
+    } Data;
+  };
+
+When :option:`EnableCountingEnumHeuristic` is true:
```
```suggestion
When :option:`EnableCountingEnumHeuristic` is `true`:
```

EugeneZelenko commented on clang-tools-extra/docs/clang-tidy/checks/bugprone/tagged-union-member-count.rst:
```diff
@@ -0,0 +1,275 @@
+separated strings that are used to search for possible counting enum constants.
+These strings are matched case insensitively as prefixes and suffixes
+respectively on the names of the enum constants.
+If :option:`EnableCountingEnumHeuristic` is false then these options do nothing.
```
```suggestion
If :option:`EnableCountingEnumHeuristic` is `false` then these options do nothing.
```

EugeneZelenko commented on clang-tools-extra/docs/clang-tidy/checks/bugprone/tagged-union-member-count.rst:
```diff
@@ -0,0 +1,275 @@
+respectively on the names of the enum constants.
+If :option:`EnableCountingEnumHeuristic` is false then these options do nothing.
+
+The default value of CountingEnumSuffixes is "count" and of
```
```suggestion
The default value of :option:`CountingEnumSuffixes` is `count` and of
```

EugeneZelenko commented on clang-tools-extra/docs/clang-tidy/checks/bugprone/tagged-union-member-count.rst:
```diff
@@ -0,0 +1,275 @@
+If :option:`EnableCountingEnumHeuristic` is false then these options do nothing.
+
+The default value of CountingEnumSuffixes is "count" and of
+CountingEnumPrefixes is "" (empty string).
```
```suggestion
:option:`CountingEnumPrefixes` is empty string.
```

EugeneZelenko commented on clang-tools-extra/docs/clang-tidy/checks/bugprone/tagged-union-member-count.rst:
```diff
@@ -0,0 +1,275 @@
+The default value of CountingEnumSuffixes is "count" and of
+CountingEnumPrefixes is "" (empty string).
+
+When :option:`EnableCountingEnumHeuristic` is true and CountingEnumSuffixes
```
```suggestion
When :option:`EnableCountingEnumHeuristic` is `true` and :option:`CountingEnumSuffixes`
```

EugeneZelenko commented on clang-tools-extra/docs/clang-tidy/checks/bugprone/tagged-union-member-count.rst:
```diff
@@ -0,0 +1,275 @@
+CountingEnumPrefixes is "" (empty string).
+
+When :option:`EnableCountingEnumHeuristic` is true and CountingEnumSuffixes
+is "count;size":
```
```suggestion
is `count;size`:
```

EugeneZelenko commented on clang-tools-extra/docs/clang-tidy/checks/bugprone/tagged-union-member-count.rst:
```diff
@@ -0,0 +1,275 @@
+    } Data;
+  };
+
+When :option:`EnableCountingEnumHeuristic` is true and CountingEnumPrefixes is "maxsize;last_"
```
```suggestion
When :option:`EnableCountingEnumHeuristic` is `true` and :option:`CountingEnumPrefixes` is `maxsize;last_`:
```

EugeneZelenko commented on clang-tools-extra/docs/clang-tidy/checks/bugprone/tagged-union-member-count.rst:
```diff
@@ -0,0 +1,275 @@
+
+This option is disabled by default.
+
+When :option:`StrictMode` is false:
```
```suggestion
When :option:`StrictMode` is `false`:
```

EugeneZelenko commented on clang-tools-extra/docs/clang-tidy/checks/bugprone/tagged-union-member-count.rst:
```diff
@@ -0,0 +1,275 @@
+      } Data;
+    };
+
+When :option:`StrictMode` is true:
```
```suggestion
When :option:`StrictMode` is `true`:
```

HerrCai0907 commented on clang-tools-extra/clang-tidy/bugprone/TaggedUnionMemberCountCheck.cpp:
```diff
@@ -0,0 +1,186 @@
+
+void TaggedUnionMemberCountCheck::registerMatchers(MatchFinder *Finder) {
+
+  static const auto UnionField = fieldDecl(hasType(qualType(
+      hasCanonicalType(recordType(hasDeclaration(recordDecl(isUnion())))))));
+
+  static const auto EnumField = fieldDecl(hasType(
+      qualType(hasCanonicalType(enumType(hasDeclaration(enumDecl()))))));
+
+  static const auto hasMultipleUnionsOrEnums = anyOf(
+      fieldCountOfKindIsGT(UnionField, 1), fieldCountOfKindIsGT(EnumField, 1));
```
Do we really need to use static const variable here? It looks like not a common pattern in `registerMatchers`.

HerrCai0907 commented on clang-tools-extra/clang-tidy/bugprone/TaggedUnionMemberCountCheck.cpp:
```diff
@@ -0,0 +1,186 @@
+               ast_matchers::internal::Matcher<FieldDecl>, InnerMatcher,
+               unsigned, N) {
+  unsigned MatchCount = 0;
+  for (const auto Field : Node.fields()) {
+    if (InnerMatcher.matches(*Field, Finder, Builder)) {
+      MatchCount += 1;
+    }
+  }
```
`llvm::count_if` can simplify logic.
Or at least remove the brace since there is only one statement here.

HerrCai0907 reviewed (approved):
In general LGTM.
Please fix some nits

tigbr commented on clang-tools-extra/clang-tidy/bugprone/TaggedUnionMemberCountCheck.cpp:
```diff
@@ -0,0 +1,186 @@
+  }
+
+  if (EnableCountingEnumHeuristic && LastEnumConstant &&
+      isCountingEnumLikeName(LastEnumConstant->getName()) &&
+      (LastEnumConstant->getInitVal() == (EnumValues.size() - 1))) {
```
I don't think I fully understand what you mean, could you show me a code example? With that said, I would rather add support for such cases in a follow-up PR.

HerrCai0907 replied:
e.g. use this kind of enum as tag

```c++
enum A {
  a = 256,
  b = 257,
};
```

tigbr replied:
I see what you mean but I feel like this would be a crude heuristic. Using an enum counter this way is misses the point in my opinion.

If you don't mind I would prefer to leave this feature out of the current PR.

tigbr committed:
Fixed nits in documentation and implementation.

tigbr committed:
Merge branch 'main' into tagged-union-member-count

Szelethus reviewed (approved):
Since the PR recieved two accepts from well-established tidy contributors, and from what I can see all comments have been adequetly answered, I say we should close this out!

LGTM, well done!
